contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,355 @@
+# Scheduling Policy
+
+_Status: Draft_
+_Author: @arnaudmz, @yastij_
+_Reviewers: @bsalamat, @liggitt_
+
+# Objectives
+
+-  Define the concept of scheduling policies
+-  Propose their initial design and scope
+
+## Non-Goals
+
+-  How taints / tolerations work
+-  How NodeSelector works
+-  How node / pod affinity / anti-affinity rules work
+-  How several schedulers can be used within a single cluster
+-  How priority classes work
+
+# Background
+
+During real-life Kubernetes architecting we encountered contexts where role-isolation (between administration and simple namespace usage in a multi-tenant context) could be improved. So far, no restriction is possible on toleration, priority class usage, nodeSelector, anti-affinity depending on user permissions (RBAC).
+
+Identified use-cases aim to ensure that administrators have a way to restrict users or namepace when
+-  using schedulers,
+-  placing pods on specific nodes (master roles for instance),
+-  using specific priority classes,
+-  expressing pod affinity or anti-affinity rules.
+
+# Overview
+
+Implementing SchedulingPolicy implies:
+-  Creating a new resource named **SchedulingPolicy** (schedpol)
+-  Creating an **AdmissionController** that dehaves on a deny-all-but basis
+-  Allow SchedulingPolicy to be used by pods using RoleBindings or ClusterRoleBindings
+
+# Detailed Design
+
+SchedulingPolicy resources are supposed to apply in a deny-all-except approach. They are designed to apply in an additive way (i.e and'ed). From Pod's perspective, a pod can use one or N of the allowed items.
+
+An AdmissionController must be added to the validating phase and must reject pod scheduling if the serviceaccount running the pod is not allowed to specify requested NodeSelectors, Scheduler-Name, Anti-Affinity rules, Priority class, and Tolerations.
+
+All usable scheduling policies (allowed by RBAC) are merged before evaluating if scheduling constraints defined in pods are allowed.
```

Comment:
Clarify what "merged" means. That seems potentially problematic, especially in case of computing coverage of conflicting scheduling components (policy A allowed this toleration, policy B allowed that toleration, policy C required nodeSelector component master=false, policy D allows nodeSelector component master=true, etc)

Reply:
As long as there was no required components nor default value, merging was quite trivial, but given that need, I guess we'll have to work on it.

I'm thinking of some ways like:
-  allowed-like rules keep having an additive behaviour:
if policy A allows nodeSelector key=a and policy B allows nodeSelector key=b => the merge produces nodeSelector key can be in [a,b]

- require-like rules prevent over allowed values => allowed values not present in required values are eventually not allowed

-  for default-like and required-like conditions, we could consider either to weight policies or to sort policies (by name?) and apply a last-seen wins rule.

Any thoughts?

Reply:
@arnaudmz @liggitt - to me, if a user specifies a required nodeselector `master=trueˋ and added a default nodeselector `master=falseˋunder another policy, I’d expect the required to superseed the default one.

Reply:
prefer to have `allow`, `deny` and `ignore` concept; so the result will be no `deny` term for the requirements.

- `allow`: passed by policy
- `deny`: rejected by policy
- `ignore`: the policy did not include term for it.

> if policy A allows nodeSelector key=a and policy B allows nodeSelector key=b => the merge produces nodeSelector key can be in [a,b]

That's error-prone and complex; especially some corner cases.

Reply:
@k82cn - I'd go with `require`, `allow`, `default` and `ignore`

Reply:
I mean the the single term; so the request is passed only all term passed :)